### PR TITLE
Auto steps per epoch

### PR DESCRIPTION
**Pull request by huffmanjohnf**
## What does this PR do?
We were only setting the learning rate scheduler `steps_per_epoch` attribute manually based on hand-calculated values, but this PR enables the attribute to be set automatically.

### CODE REVIEW

File: flower_classifier/datasets/random.py [hunk 1]
```diff
@@ -23,6 +23,7 @@ def __init__(self, batch_size=16):
         self.batch_size = batch_size
         self.train_dataset = RandomDataset(n_examples=64)
         self.val_dataset = RandomDataset(n_examples=32)
+        self.dims = (64,)
```
Comment (jeremyjordan) [inline flower_classifier/datasets/random.py]: we should instead set this attribute to `(len(train_dataset), len(val_dataset))` to include both sizes and avoid hardcoding a number

File: flower_classifier/datasets/oxford_flowers.py [hunk 2]
```diff
@@ -185,11 +185,12 @@ def __init__(self, data_dir=ROOT_DATA_DIR, batch_size=64):
         self.data_dir = data_dir
         self.batch_size = batch_size
 
-    def setup(self, stage=None):
+    def prepare_data(self):
         self.dataset = OxfordFlowers102Dataset(self.data_dir, transforms=DEFAULT_IMG_TRANSFORMS)
         train_idx, valid_idx = self.get_sampler_indices()
         self.train_sampler = SubsetRandomSampler(train_idx)
         self.val_sampler = SubsetRandomSampler(valid_idx)
+        self.dims = (len(train_idx),)
```
Comment (jeremyjordan) [inline flower_classifier/datasets/oxford_flowers.py]: let's also include the validation dataset size for completeness

File: flower_classifier/train.py [hunk 3]
```diff
@@ -14,6 +15,14 @@
 @hydra.main(config_path="../conf", config_name="config")
 def train(cfg):
     data_module = hydra.utils.instantiate(cfg.dataset)
+    data_module.prepare_data()
```
Comment (jeremyjordan) [inline flower_classifier/train.py:30]: i don't love the idea of reaching in to manually invoke this, but can't think of a better option at the moment

File: flower_classifier/train.py [hunk 4]
```diff
@@ -14,11 +15,16 @@
 @hydra.main(config_path="../conf", config_name="config")
 def train(cfg):
     data_module = hydra.utils.instantiate(cfg.dataset)
+
+    lr_scheduler = getattr(cfg, "lr_scheduler", None)
+    scheduler = getattr(lr_scheduler, "scheduler", None)
+    steps_per_epoch = getattr(scheduler, "steps_per_epoch", None)
+    if steps_per_epoch == "AUTO":
+        data_module.prepare_data()
+        lr_scheduler.scheduler.steps_per_epoch = math.ceil(data_module.len_train / cfg.dataset.batch_size)
```
Comment (jeremyjordan) [inline flower_classifier/train.py]: let's wrap this into a function which accepts the `DictConfig` object and returned a modified version, that way we can write a couple tests to ensure this works as we intend.

**Review by jeremyjordan — APPROVED** [review]
👍
